src/com/perl5/lang/perl/idea/refactoring/extract/PerlExtractMethodHandler.java

```diff
@@ -0,0 +1,139 @@
+import com.intellij.openapi.editor.SelectionModel;
+import com.intellij.openapi.project.Project;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.refactoring.RefactoringActionHandler;
+import com.intellij.refactoring.extractMethod.AbstractExtractMethodDialog;
+import com.intellij.refactoring.extractMethod.ExtractMethodDecorator;
+import com.intellij.refactoring.extractMethod.ExtractMethodValidator;
+import com.intellij.refactoring.util.AbstractVariableData;
+import com.intellij.util.containers.ArrayListSet;
+import com.perl5.lang.perl.fileTypes.PerlFileTypeScript;
+import com.perl5.lang.perl.idea.refactoring.PerlRefactorUtils;
+import com.perl5.lang.perl.idea.refactoring.PerlRefactor;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import java.util.Set;
+
+public class PerlExtractMethodHandler implements RefactoringActionHandler {
+
+  @Override
+  public void invoke(@NotNull Project project, Editor editor, PsiFile file, DataContext dataContext) {
+
+    final SelectionModel selectionModel = editor.getSelectionModel();
+    if (!selectionModel.hasSelection()) selectionModel.selectLineAtCaret();
+
+    final int selectionStartLine = selectionModel.getSelectionStart();
+    final int selectionLength = selectionModel.getSelectionEnd() - selectionStartLine;
+    String selectedText = selectionModel.getSelectedText();
+
+    if (selectedText == null || selectedText.isEmpty()) {
+      return;
+    }
+
+    PerlRefactor perlRefactor = new PerlRefactor(selectedText, "myNewMethod");
+
+    Set<String> parameterVariables = new ArrayListSet<>();
+    parameterVariables.addAll(perlRefactor.getParameters());
+
+    Set<String> returnVariables = new ArrayListSet<>();
+    returnVariables.addAll(perlRefactor.getInnerVariables());
+
+    // TODO: create PerlCodeFragment that contains my(...) = foo(bar);
+    // TODO: figure how to use returnInside
+    CodeFragment fragment = new CodeFragment(parameterVariables, returnVariables,false);
+
+    // This populates the parameters and signature
+    final ExtractMethodDecorator decorator = (methodName, variableData) -> {
+
+      StringBuilder builder = new StringBuilder();
+      boolean first = true;
+
+      if (perlRefactor.getOuterVariablesModified().size() > 0) {
+        builder.append("my (");
+        for (String ov : perlRefactor.getOuterVariablesModified()) {
+          if (first) {
+            first = false;
+          }
+          else {
+            builder.append(", ");
+          }
+          builder.append(ov);
+        }
+        builder.append(") = ");
+        first = true;
+      }
+
+      builder.append(methodName).append("(");
+      for (AbstractVariableData vd : variableData) {
+        if (!vd.passAsParameter) continue;
+        if (first) {
+          first = false;
+        }
+        else {
+          builder.append(", ");
+        }
+        builder.append(vd.name);
+      }
+      builder.append(");");
```

Comment:
You should build `my ($var1, ....) = @_` based on `AbstractVariableData`. Your approach won work.